### ISSUE TITLE
Section 7.1, "Registering a platform OpenAPI", should not specify implementation detail

### ISSUE THREAD

**Issue by msmiths** [eclipse/microprofile-open-api#443]
### Overview

[Section 7.1, "Registering a platform OpenAPI"](https://github.com/eclipse/microprofile-open-api/blob/master/spec/src/main/asciidoc/microprofile-openapi-spec.adoc#71-registrying-a-platform-openapi), of the MP Open specification states the value of the `mp.openapi.spi.platform.<component>` MP Config property is the location of the  component’s OAS3 document.  The example given suggests that this is the location of the component’s OAS3 document on disk.  However, the specification also states that this is an SPI that is only intended to be set by the vendor framework and is not intended to be used by user applications.  Given this, the implementation should be left to the vendor since it is not visible to end users.  This section of the specification should be rewritten to remove any implementation details.

**Comment by arthurdm** [comment]
hi @msmiths - from what I recall (it was a while ago 😄 ) the purpose of specifying exactly the MP Config key and the semantics of its corresponding value was to bring consistency to the way _any_ MP OpenAPI compliant implementation could be extended to expose REST APIs of the framework - separately from the REST APIs of the applications.

I think what may be throwing off the understanding of this spec section is the different targeted persona:  if we think there's an "ops persona" and a "app owner persona", this spec section is providing guidance for the former, not the latter.  In many cases the two roles can be done by the same person, but that is not a requirement.

So in summary - it's definitely an option to remove the ops persona guidance and leave it as a black box for implementations to figure out, but that would cost the consistency of the programming model for that persona.

**Comment by msmiths** [comment]
@arthurdm I understand your argument, but I respectfully disagree. The important aspect of this functionality is the externals, i.e., the behaviour that an end user can expect from the `/openapi/platform` and `/openapi/platform/<component>` endpoints. How each vendor implements this functionality under the covers is not important to the end user. Indeed, given that each vendors runtime is likely to be very different in terms of the implementation, trying to impose a consistent programming model is likely to make it more difficult for some vendors to implement this functionality than others.

**Comment by arthurdm** [comment]
> The important aspect of this functionality is the externals, i.e., the behaviour that an end user can expect 

I think this hits exactly the point - it's about which persona we're talking about.  If the "end user" we're talking about is the person who is just _using_ the app server (or MP impl), then I completely agree, this persona does not care how `/openapi/platform` is setup  underneath the covers.  

But if the "end user" we're talking about is the person who is _setting up / configuring_ the app server (or MP impl), the person may want to augment that core app server functionality to expose additional endpoints via `/openapi/platform` (e.g. a stack product API).

**Comment by msmiths** [comment]
But that is not what the specification says.  The first line of section 7.1 states:

    The components within a vendor’s framework will register their platform OpenAPI by using the following pattern:

This implies that the act of registering the OpenAPI document for a component is not performed by an operator, it is an automated task performed by the component.  In OL, we are looking to implement this functionality by updating any component that wants to expose an OpenAPI document so that they set the relevant `mp.openapi.spi.platform.<component>` property as a system property when the bundle is activated. No action is required on the part of the operator.

**Comment by arthurdm** [comment]
That's a fair point @msmiths - that part is specifically talking about components. 

I think if we augmented the spec to say that the value of this property could also be the OAS3 doc itself (as a String), that would help in terms of flexibility?

**Comment by msmiths** [comment]
@arthurdm Yes, this would help.  I will make this change to the spec, along with making support for section 7 of the spec optional, as discussed in the MP OpenAPI hangout yesterday.

**Comment by msmiths** [comment]
@EricWittmann During further discussions with @arthurdm , I made the following points:

- The OpenAPI Specification (OAS) already provides a standards based mechanism for documenting REST APIs, so the relevant section of the MP OpenAPI 2.0 specification is not adding anything here.
- Other MP components that expose REST APIs (MP Metrics, MP Health, etc.) could, and arguably should, provide an OpenAPI document that documents their REST APIs and this document should be available in GitHub along with the source code, specification, etc., for that component.
- Other non-MP components in vendor runtimes that expose REST APIs could also provide an OpenAPI document that documents their REST APIs.  These documents could be made available as part of the documentation for the relevant component.
- This leaves the primary use case for the functionality as being a mechanism to "discover" features provided by the MP platform but, as you alluded to above, this is not a particularly compelling argument for including this functionality in the specification.

Given the points above, I believe that the best course of action at this point in time is to remove this section from the specification because it does not appear to be addressing any immediate problems experienced by consumers of MP OpenAPI.

However, I do believe that the general approach can be reinstated at some point in the future when we finally come to address the problem of hosting multiple applications that expose REST APIs.  In this scenario, it could be used to provide a mechanism that allows users to discover the applications that are available in a given runtime and then allow them to retrieve the OpenAPI document for a specific application.  This would avoid the need to merge the OpenAPI documents for multiple applications together into a single OpenAPI document, with all of the associated problems that this causes.  This is analogous to the service document that is exposed when using the Atom Publishing Protocol (APP).  The MP OpenAPI "service document" could then be extended to contain references to components in the runtime that expose REST APIs... not just applications.

Thoughts?

**Comment by phillip-kruger** [comment]
+1

**Comment by EricWittmann** [comment]
Yeah I think I agree with all of that.  Unless there's a use-case I'm not seeing...

**Comment by Emily-Jiang** [comment]
+1! It can't be released without TCKs anyway.

**Comment by msmiths** [comment]
Section 7 has been removed from the upcoming version of the specification (2.0). Closing this issue.

**Comment by msmiths** [comment]
Section 7 has been removed from the upcoming version of the specification (2.0). Closing this issue.
